### ISSUE TITLE
High input lag

### ISSUE THREAD

**Issue by Tasosgemah** [libretro-mirrors/beetle-saturn-libretro#36]
The core seems to have higher input lag compared to other cores. That's from my own testing. Using the frames advance method it seems like all games have at least 5 frames of lag, as opposed to other cores that have 3 or less (GPU sync is ON and everything is tested with a CRT PC monitor).

Games like Sonic Jam (the classic Sonic games within) are literally unplayable as the delay is too much for a game of this kind. In comparison, Sonic on Genesis GX plus only has 1 frame of lag.

There is also a big difference between a real Saturn and Mednafen. This difference is smaller between other cores and the real systems they emulate.

**Comment by Ryunam** [comment]
Yeah, the input latency with this core is abnormally high and very noticeable, to the extent that it makes the vast majority of games close to unplayable.
Maybe @Brunnis could take a look at this? His work on reducing this issue for NES and SNES cores was excellent.

**Comment by Brunnis** [comment]
I'm sorry, but I really can't look into this at the moment. I have my hands completely full and no spare time left to spend on this. Besides, I think this particular issue is something the original author(s) should look into (since it isn't likely to be libretro related). Such input lag issues should really be at the top of their todo list, since it has such a direct effect on gameplay quality.

**Comment by ghost** [comment]
is anyone able to re-test this using latest commits if theres any change?

**Comment by Tasosgemah** [comment]
Just tested the latest version i could download as of now. There is no difference, it's still laggy. I test Duke Nukem and it takes 8 frames to shoot and 9 frames to jump on a CRT monitor and a wired controller. Just as laggy as the core ever was from the start. Was there supposed to be an improvement lately?

**Comment by ghost** [comment]
not sure if there is any. are you able to compare with other saturn emu?

**Comment by Tasosgemah** [comment]
It's been a while since i used SSF but i remember it having high input lag as well. I can't test Yabause on RetroArch because the frame advance acts weird with it plus it has too many glitches. It's impossible to test Sonic 1 for instance with it.

**Comment by Ryunam** [comment]
Is this same amount of input lag also present in the standalone version of Mednafen Saturn?

**Comment by ghost** [comment]
@Tasosgemah ill repeat my reply to you last Nov 2017: "are you able to compare with other saturn emu?" or even tested in standalone? if this happens similar in standalone then nothing can do about that until upstream updates too since core is constantly WIP. each version since 0.9.41 is mostly for SS related.

**Comment by hiddenasbestos** [comment]
Personally I've not felt lag to be present in either libretro or mednafen in games such as Megaman X4 or Street Fighter Alpha 2.

I'm running a stock i7-3770. Presumably it doesn't degrade well with slower computers (not that I feel I have a beast of a PC or anything) or perhaps there's a different reason for the lag?

**Comment by cooperate** [comment]
Can we get this integrated as standard in the core?

Set ss.midsync to 1 in mednafen-09x.cfg (will require much more CPU)

Reference thread:
https://www.reddit.com/r/emulation/comments/71dfkf/saturn_reduce_input_lag_in_mednafen/

**Comment by Tasosgemah** [comment]
I have an i5 4670 and the core runs at full speed with Hard GPU sync ON (with zero GPU frames) and mid-frame sync ON. But i still see the same input lag as always. Sonic in Sonic Jam games reacts on the 5th frame (on Genesis he reacts on the 2nd) and Duke Nukem reacts on the 8th frame which does feel slower than the real machine (can't measure the real one though so i can't prove it). 

I also measured Megaman X4 which was mentioned with frame advance and that seems faster, Megaman reacts on the 4th frame. Same with Fighter's Megamix which also reacts on the 4th frame.

Which means that, so far, and until a game that reacts faster is found, it seems like there's always a 3 frame input lag on the Beetle Saturn core, plus whatever each game adds. Which makes it the slowest to react core.

Unfortunately, i can't measure the standalone Mednafen, or any other standalone emulator, since they don't have a frame advance feature. So i don't know how it compares with the libretro core.

**Comment by hiddenasbestos** [comment]
As annoying as this sounds, to fix this properly it really needs to be measured on real Saturn hardware. It could be perfectly accurate already or actually off by some other number of frames. I'm not saying there's not a problem but it needs actual evidence. Right now all we have are theories and niggles.

Using frame advance to measure lag, I assume, should remove errors caused by input and GPU latency. It suggests to me that it really does take 5 frames for Sonic to jump on Saturn (or at least on this emulator). I've measured the game myself and I did find the same number of delay frames on my system in RetroArch. 

There are possible reasons for this - it could be a genuine emulation error (e.g. the game code is always being sent the previous frames input data - which would add +1 delay) but it could also have been sloppy coding or a deliberate design change at SEGA. It could be a combination of several different factors.

Finally don't forget this core is a port of the upstream Mednafen code so any errors should probably be fixed there (although RetroArch frame advance does provide an advantage in measuring the delay frames). Again some measurement of frame delay on Mednafen would be good -- perhaps record it in OBS with the Game controllers control dialog open so you can see when the button lights up and then frame advance the video to see when Sonic jumps (measuring RetroArch the same way as a control).

**Comment by Tasosgemah** [comment]
Yeah i understand that someone has to measure a real Saturn and provide evidence. And although i do have a Saturn i don't know how to do this, nor have any equipment.

**Comment by Boviligg** [comment]
I just tested standalone mednafen with the ol' frame advance method and it has the same amount of internal frames of lag as the RA core with hardgpu sync set to 0. 
The game tested was Shippu Mahou Daisakusen - It takes 4 frames until you shoot your weapon and 5 frames until you move your ship.
